### PR TITLE
MF-798 - Add utf8 support for email validation

### LINKED ISSUE #798: Length of an email validation
The email validation introduced with #795 has some problems.

According to the [rfc3696 section 3.](https://tools.ietf.org/html/rfc3696#section-3) the maximum length of an email address is A@B A(local part)=64 characters and B(domain part)=255 characters.
The minimal length is three because a valid email address has the form "a@b" and the smallest length for a and b is one.

The email user (the local part) and the host (the domain part) can also have non ASCII UTF-8 characters. These regex expressions will make them invalid.
Do we want to enable email addresses which have UTF-8 characters as well?

@mainflux/contributors any thoughts on this?

### PR DESCRIPTION
**Pull request by filinvadim**
### What does this do?
Fix email address segments legth validation.
Adds validation of email addresses with utf8 encoded characters.

### Which issue(s) does this PR fix/relate to?
Resolves #798 .

### List any changes that modify/break current functionality
External dependency: `golang.org/x/net/idna`.

### Have you included tests for your changes?
Yes. Tests are extended.

### Did you document any new/modified functionality?
No.

### Notes
-

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -167,8 +167,6 @@ github.com/mainflux/mainflux v0.0.0-20191223163044-f42f2095bab4/go.mod h1:K3ghSI
 github.com/mainflux/mainflux v0.0.0-20200314190902-c91fe0d45353/go.mod h1:yijZGLNkcDOPJfPhRMwuu5ZFcNHqDHzWurN4q1rOT/Q=
 github.com/mainflux/mproxy v0.1.3 h1:/JNnxgo/03wSpbwQH2+WE1AzgMWaSKogTVXblh18x5s=
 github.com/mainflux/mproxy v0.1.3/go.mod h1:/BdaBfgye1GNCD+eat4ipFamy9IEVRH5nhZS0yEShVg=
-github.com/mainflux/mproxy v0.1.4 h1:g7LhJgA+BXlgaZeyLj505aD5DCsrix79TEiLWuHJZd4=
-github.com/mainflux/mproxy v0.1.4/go.mod h1:MBLtv/RvhT8QsmXz4g3GxkRaP8PqlVqBWeqvw9QmO8k=
```
Comment (manuio) [inline go.sum]: Is this necessary?

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users.go [hunk 2]
```diff
@@ -9,9 +9,14 @@ import (
...
 
 	"github.com/mainflux/mainflux/errors"
+	"golang.org/x/net/idna"
 )
 
-const minPassLen = 8
+const (
+	minPassLen = 8
+	minLocalLen = 64
+	minHostLen = 255
```
Comment (manuio) [inline users/users.go]: min or max?

**Review by manuio — CHANGES_REQUESTED** [review]

File: go.sum (hunk 1 above)
Comment (filinvadim) [reply go.sum]: Not sure. Go mod tidy did this. What should I do? Revert?

File: users/users.go [hunk 3]
```diff
@@ -57,7 +62,7 @@ type UserRepository interface {
 }
 
 func isEmail(email string) bool {
-	if len(email) < 6 || len(email) > 254 {
+	if len(email) == 0 {
```
[the remark this answers is not in the document]
Comment (filinvadim) [reply users/users.go]: ok

File: users/users.go (hunk 2 above)
Comment (filinvadim) [reply users/users.go]: thanks! typo

File: go.sum (hunk 1 above)
Comment (manuio) [reply go.sum]: Yes please

Comment (filinvadim) [reply go.sum]: Done

File: users/users.go [hunk 4]
```diff
@@ -66,14 +71,23 @@ func isEmail(email string) bool {
 		return false
 	}
 
-	user := email[:at]
+	local := email[:at]
 	host := email[at+1:]
```
Comment (manuio) [inline users/users.go]: Split would simplify all here:
```
    es := strings.Split(email, "@")
    local, host := es[0], es[1]
```

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users.go [hunk 5]
```diff
@@ -66,14 +71,23 @@ func isEmail(email string) bool {
...
 
-	if len(user) > 64 {
+	if len(local) > maxLocalLen || len(host) > maxHostLen {
```
Comment (manuio) [inline users/users.go]: You could then check here `if len(local) == 0 || ...`

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users.go (hunk 4 above)
Comment (filinvadim) [reply users/users.go]: I've tried to make minimum changes

Comment (filinvadim) [reply users/users.go]: It is out of issue scope

File: users/users.go (hunk 5 above)
Comment (filinvadim) [reply users/users.go]: ok

File: users/users.go (hunk 4 above)
Comment (manuio) [reply users/users.go]: I understand and I know that you didn't write this code but it would be the perfect moment to fix this :)

**Review by manuio — APPROVED** [review]
LGTM!

File: users/users.go [hunk 6]
```diff
@@ -57,23 +62,34 @@ type UserRepository interface {
...
+
+	es := strings.Split(email, "@")
+	if len(es) != 2 {
 		return false
 	}
+	local, host := es[0], es[1]
 
-	at := strings.LastIndex(email, "@")
-	if at <= 0 || at > len(email)-3 {
+	if len(local) == 0 || len(host) == 0 {
```
Comment (manuio) [inline users/users.go]: Wait, we are not checking the extension anymore. Or you revert this `len(host) - 3 == 0` or you can maybe do an extra Split to know the real size of the extension.

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users.go [hunk 7]
```diff
@@ -57,23 +62,44 @@ type UserRepository interface {
...
 
-	user := email[:at]
-	host := email[at+1:]
+	if len(local) == 0 {
+		return false
+	}
 
-	if len(user) > 64 {
+	hs := strings.Split(host, ".")
+	if len(es) != 2 {
```
Comment (manuio) [inline users/users.go]: Fix `len(hs)`

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users.go [hunk 8]
```diff
@@ -57,23 +62,44 @@ type UserRepository interface {
...
+		return false
+	}
+	domain, ext := hs[0], hs[1]
+	
+	if len(domain) == 0 || len(ext) == 0 {
+		return false
+	}
+
+	if len(local) > maxLocalLen || len(host) > maxHostLen {
```
Comment (manuio) [inline users/users.go]: Please, add here a `maxExtensionLen` check

**Review by manuio — CHANGES_REQUESTED** [review]

File: users/users_test.go [hunk 9]
```diff
@@ -51,6 +51,20 @@ func TestValidate(t *testing.T) {
...
+			},
+			err: nil,
+		},
+		"validate user with utf8 email (hieroglyph)": {
+			user: users.User{
+				Email:    "艾付忧西开@艾付忧西开.再得",
+				Password: password,
+			},
+			err: nil,
+		},
```
Comment (manuio) [inline users/users_test.go:82]: Thank you for your help! If you have time and want to make happy our codecover you could add here 3 tests for invalid local, invalid domain and invalid extension.

File: users/users.go [hunk 10]
```diff
@@ -57,23 +63,48 @@ type UserRepository interface {
...
-	if at <= 0 || at > len(email)-3 {
+	es := strings.Split(email, "@")
+	if len(es) != 2 {
 		return false
 	}
+	local, host := es[0], es[1]
 
-	user := email[:at]
-	host := email[at+1:]
+	if len(local) == 0 {
```
Comment (manuio) [inline users/users.go]: I would do `if len(local) == 0 || len(local) > maxLocalLen{` here

File: users/users.go [hunk 11]
```diff
@@ -57,23 +63,48 @@ type UserRepository interface {
...
+		return false
+	}
+
+	hs := strings.Split(host, ".")
+	if len(hs) != 2 {
+		return false
+	}
+	domain, ext := hs[0], hs[1]
+
+	if len(domain) == 0 || len(ext) == 0 {
```
Comment (manuio) [inline users/users.go]: and here ` if len(domain) == 0 || len(domain) > maxDomainLen`

File: users/users.go [hunk 12]
```diff
@@ -57,23 +63,48 @@ type UserRepository interface {
...
+		return false
+	}
 
-	if len(user) > 64 {
+	if len(local) > maxLocalLen || len(host) > maxHostLen {
+		return false
+	}
+
+	if len(ext) > maxTLDLen {
```
Comment (manuio) [inline users/users.go]: and finally here `if len(ext) == 0 || len(ext) > maxTLDLen`

**Comment by filinvadim** [comment]
Let's stop here.

**Comment by drasko** [comment]
@filinvadim how should I understand this comment? Maybe you wanted to say "Thanks @manuio for taking your time in reviewing my PR, I understand that Mainflux is high-quality project that demands though peer-reviews and skills and patience are needed to get my PR to the adequate level to be accepted"?

**Comment by filinvadim** [comment]
@drasko i did several out of issue scope changes and then I stopped. 
@manuio thanks for taking your time in reviewing my PR! You’re great!
Feel free to merge it or not.

**Comment by drasko** [comment]
@filinvadim of course it will not be merged, not until all remarks are resolved. And this is only one reviewer/maintainter, there are several of us.

I am closing the issue, I hope this also clarifies why you and other in the industry can trust Mainflux code.

File: users/users_test.go [hunk 13]
```diff
@@ -51,6 +51,64 @@ func TestValidate(t *testing.T) {
...
+		"validate user with no email tld": {
+			user: users.User{
+				Email:    "user@example.",
+				Password: password,
+			},
+			err: users.ErrMalformedEntity,
+		},
+		"validate user with too long email tld": {
+			user: users.User{
+				Email:    "user@example.aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
```
Comment (manuio) [inline users/users_test.go]: This and specially the next one are too long. You need a function here. We already did in other places things like:
```
letters = "abcdefghijklmnopqrstuvwxyz"
func randomString(n int) string {
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}

longDomain := randomString(maxDomainLen + 1)
```

**Review by manuio — CHANGES_REQUESTED** [review]

**Review by manuio — APPROVED** [review]
LGTM!

**Review by drasko — APPROVED** [review]
LGTM

**Comment by manuio** [comment]
Thank you @filinvadim !

**Pull request merged**